session.go

```diff
@@ -469,42 +417,50 @@ func (s *Session) sendLoop() error {
 		default:
 		}
 
+		// Flushes at least once every 100 microseconds unless we're
+		// constantly writing.
+		var buf []byte
 		select {
-		case ready := <-s.sendCh:
-			// Commit to perform the write, iff it has not expired prior to being consumed from the ch.
-			if !atomic.CompareAndSwapUint32(&ready.Stage, stageInitial, stageFinal) {
-				continue
-			}
-
-			// Send a header if ready
-			if ready.Hdr != nil {
-				sent := 0
-				for sent < len(ready.Hdr) {
-					n, err := s.conn.Write(ready.Hdr[sent:])
-					if err != nil {
-						s.logger.Printf("[ERR] yamux: Failed to write header: %v", err)
-						asyncSendErr(ready.Err, err)
-						return err
+		case buf = <-s.sendCh:
+		default:
+			select {
+			case buf = <-s.sendCh:
+			case <-s.shutdownCh:
+				return nil
+			case <-writeTimeout.C:
+				if err := writer.Flush(); err != nil {
+					if isTimeout(err) {
+						err = ErrConnectionWriteTimeout
 					}
-					sent += n
+					return err
 				}
-			}
+				returnBuffer(writer)
+				writer = nil
 
-			// Send data from a body if given
-			if ready.Body != nil {
-				_, err := io.Copy(s.conn, ready.Body)
-				if err != nil {
-					s.logger.Printf("[ERR] yamux: Failed to write body: %v", err)
-					asyncSendErr(ready.Err, err)
-					return err
+				select {
+				case buf = <-s.sendCh:
+				case <-s.shutdownCh:
+					return nil
 				}
+
+				writer = getBuffer(s.conn)
+				writeTimeout.Reset(100 * time.Microsecond)
```

Comment:
shouldn't we `Stop` the timer and possibly drain the channel before resetting?

Reply:
The timer has to have fired for us to get to this point.